feat: ucan endpoint supports agent message

vasco-santos opened the pull request:
As agreed, first iteration adds support for `POST /ucan` into the API. `upload-api` is still using old world, therefore we need to keep both `ucanto` dependencies for this first iteration. Code changes aim to just drop most of the functions in place in favour of the added ones once in next PR we update ucanto in `upload-api`.

Decisions:
- new bucket structures based on https://github.com/web3-storage/w3infra/issues/186#issuecomment-1502309162
- we install `"@ucanto/core-next": "npm:@ucanto/core@^7.0.1"` + `"@ucanto/transport-legacy": "npm:@ucanto/transport@^5.1.1"`
  - initial goal here was to install new ones and keep olds with suffix `-legacy` to easily drop them in next PR. Sadly, given other `ucanto` packages depend on the 5th line of core, I could not work around having `"@ucanto/core-legacy": "npm:@ucanto/core@^5.1.0"` because dependencies would struggle to find correct one...
  - path forward is still relatively easy given `ucanto/core` is barely used
- tried to avoid naming workflow everywhere, moving to agentMessage to uniform with ucanto

Closes https://github.com/web3-storage/w3infra/issues/186

Gozala commented on upload-api/ucan-invocation.js:
```diff
@@ -56,6 +71,176 @@ export async function processUcanLogRequest (request, ctx) {
+    }
+
+    // @ts-expect-error headers are not needed
+    const agentMessage = await CAR.request.decode({ body: agentMessageBytes })
```
Nit: you can just pass `headers: {}` instead.

Gozala commented on upload-api/ucan-invocation.js:
```diff
@@ -1,5 +1,6 @@
 import * as CAR from '@ucanto/transport/car'
-import * as CBOR from '@ucanto/transport/cbor'
+import * as CAR_LEGACY from '@ucanto/transport-legacy/car'
+import * as CBOR_LEGACY from '@ucanto/transport-legacy/cbor'
```
BTW `@ucanto/core` exports `CBOR` codec now.

Gozala reviewed (approved):
Awesome! Thanks for getting it all done so quickly! It all looks good although I made some minor suggestions here and there.

I do however find naming misleading however, because through out `agentMessage` is used to refer to the CAR file containing an `AgentMessage` block in a root. I would suggest update naming to call it `agentArchive` or `agentMessageArchive` instead to disambiguate CAR containing the message from the message block itself.

vasco-santos commented on docs/ucan-invocation-stream.md:
```diff
@@ -29,11 +30,11 @@ Note that at the time of writing Event Archival flow is still to be implemented.
 UCAN Invocation Stack contains 3 buckets so that it can keep an audit of the entire system, while allowing this information to be queried in multiple fashions.
 
-Firstly, the **`workflow-store` bucket** stores the entire encoded file containing one or more invocations to be executed. It is stored as received from UCAN services interacting with UCAN Invocation Stream. It is keyed as `${workflow.cid}/${workflow.cid}` and its value is likely in CAR format. However, CID codec should tell if it is something else.
+Firstly, the **`workflow-store` bucket** stores the entire encoded agent message files containing invocations to be executed, and/or created receipts for ran invocations. It is stored as received from UCAN services interacting with UCAN Invocation Stream. It is keyed as `${agentMessage.cid}/${agentMessage.cid}` and its value is likely in CAR format. However, CID codec should tell if it is somethi ...
```
Per suggestion in review text, I changed this into `agentMessageArchive.cid`. Let me know if it is more clear this way

vasco-santos commented:
@Gozala addressed all your review, looks like we might need to still align on 1 point:
- https://github.com/web3-storage/w3infra/pull/187#discussion_r1163770550

Given PR is approved, I will move forward and merge this in, so that we can start working on `access-api` and `upload-api` upgrade. Happy to do any needed change in upcoming PR based on your answer there
